lib/confluence/failure_to_ship.es6.js

```diff
@@ -0,0 +1,136 @@
+          let browserKeys = browsers.map((browser) => browser.browserKey);
+          // Find all browsers that
+          //   - a previous version of the browser.
+          //   - some other browser but released in previous one year range.
+          //   - if no release in one year, uese most recent browser.
+          this.browserDAO.where(this.mlang.OR(
+            this.mlang.IN(this.Browser.BROWSER_KEY, browserKeys),
+            this.mlang.AND(
+              this.mlang.EQ(this.Browser.BROWSER_NAME, browser.browserName),
+              this.mlang.EQ(this.Browser.OS_NAME, browser.osName),
+              this.mlang.GT(this.Browser.RELEASE_DATE, previousYear),
+              this.mlang.LT(this.Browser.RELEASE_DATE, date)),
+            this.mlang.AND(
+              this.mlang.GT(this.Browser.RELEASE_DATE, previousYear),
+              this.mlang.LT(this.Browser.RELEASE_DATE, date))))
+          .select().then((result) => {
+            let browserIface = {};
+            let numOtherBrowsers = 0;
+            let majorIface = {};
+            let promises = [];
+            let prevReleaseBrowsers = [];
+            let comparedBrowsers = [];
+            for (let i = 0; i < result.a.length; i++) {
+              if (result.a[i].browserName !== browser.browserName) {
+                numOtherBrowsers++;
+                comparedBrowsers.push(result.a[i]);
+              } else if (result.a[i].browserKey !== browser.browserKey) {
+                prevReleaseBrowsers.push(result.a[i]);
+              }
+              promises.push(result.a[i].interfaces.select({
+                put: function(iface) {
+                  if (result.a[i].browserName === browser.browserName) {
+                    if (!browserIface.hasOwnProperty(iface.interfaceKey)) {
+                      browserIface[iface.interfaceKey] = true;
+                    }
+                  } else {
+                    if (!majorIface.hasOwnProperty(iface.interfaceKey)) {
+                      majorIface[iface.interfaceKey] = 0;
+                    }
+                    majorIface[iface.interfaceKey]++;
+                  }
+                },
+              }));
+            }
+            Promise.all(promises).then(() => {
+              // Find interfaces that all other vendor's browsers supports
+              // but not available in any version of this verndor's browsers.
+              let numFailureToShip = 0;
+              for (let ifaceKey in majorIface) {
+                if (majorIface[ifaceKey] !== numOtherBrowsers) continue;
+                if (!browserIface.hasOwnProperty(ifaceKey)) {
+                  numFailureToShip++;
+                }
+              }
+              this.failureToShipDAO.put(this.FailureToShipData.create(
+                {
+                  browserName: browser.browserName,
+                  browserVersion: browser.browserVersion,
+                  browser,
+                  prevReleaseBrowsers,
```

Comment:
Do we make meaningful use of all these data? Both `prevReleaseBrowsers` and `comparedBrowsers` can be recomputed given `browser`, no?